packages/apputils/src/widgettracker.ts

```diff
@@ -0,0 +1,366 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+import { InstanceTracker } from '@jupyterlab/coreutils';
+
+import { IDisposable } from '@phosphor/disposable';
+
+import { ISignal, Signal } from '@phosphor/signaling';
+
+import { FocusTracker, Widget } from '@phosphor/widgets';
+
+/**
+ * A tracker that tracks widgets.
+ *
+ * @typeparam T - The type of widget being tracked. Defaults to `Widget`.
+ */
+export interface IWidgetTracker<T extends Widget = Widget> extends IDisposable {
+  /**
+   * A signal emitted when a widget is added.
+   */
+  readonly widgetAdded: ISignal<this, T>;
+
+  /**
+   * The current widget is the most recently focused or added widget.
+   *
+   * #### Notes
+   * It is the most recently focused widget, or the most recently added
+   * widget if no widget has taken focus.
+   */
+  readonly currentWidget: T | null;
```

Comment:
I *almost* want to change this to `current`, and change `widgetUpdated` to `updated`. As long as we are expecting downstream users to change the name of the thing...

Reply:
I almost did, too. Almost.

Reply:
Especially when I realized I couldn't just pass the signal through because the `sender` argument would be incorrect. But it would make transitioning extensions over more expensive and I don't want to impose that cost.

Reply:
I'm feeling pretty tempted by this, and am curious what others think. If we decide to *not* make that API breaking change, then I'm happy with this PR as it stands.

Reply:
Unless that conversation already happened and I missed it :)

Reply:
We have not had this conversation since this PR started after the weekly meeting. I am open to it, for sure.

* I don't like those signal proxies because they are overhead.
* I don't like imposing an API change on extension authors.

Those are the two sides of the coin.

Reply:
* I don't like thinking about whether I should use `current` or `currentWidget`

Reply:
Do you think this problem goes away if we rename `InstanceTracker` to `ObjectTracker` or `Tracker`?

Reply:
It's mitigated, but there are still two `Tracker` types to keep in mind.

Reply:
Why would you have to think about whether to use `current` or `currentWidget`? Extension authors will instantiate `WidgetTracker` objects because that's what `LayoutRestorer` wants and because that's what all the core components export. I think as long as the name `InstanceTracker` disappears, the confusion dissipates.

Reply:
Yeah, I'm really just operating off of my initial assumption that `WidgetTracker` would inherit from `InstanceTracker`. That assumption is not necessarily what other developers would think, especially if they are not reviewing this PR. So I'm not too worried about it.